utility/unit_of_work/uow_leader.go

```diff
@@ -26,32 +29,58 @@ func NewLeaderUOW(height int64, readContext modules.PersistenceReadContext, rwPe
 	}
 }
 
-func (uow *leaderUtilityUnitOfWork) CreateAndApplyProposalBlock(proposer []byte, maxTxBytes uint64) (stateHash string, txs [][]byte, err error) {
-	prevBlockByzantineVals, err := uow.prevBlockByzantineValidators()
-	if err != nil {
+func (uow *leaderUtilityUnitOfWork) CreateProposalBlock(proposer []byte, maxTxBytes uint64) (stateHash string, txs [][]byte, err error) {
+	log := uow.logger.With().Fields(map[string]interface{}{
+		"proposer":   hex.EncodeToString(proposer),
+		"maxTxBytes": maxTxBytes,
+		"source":     "CreateProposalBlock",
+	}).Logger()
+	log.Debug().Msg("calling beginBlock")
+	// begin block lifecycle phase
+	if err := uow.beginBlock(); err != nil {
 		return "", nil, err
 	}
 
-	// begin block lifecycle phase
-	if err := uow.beginBlock(prevBlockByzantineVals); err != nil {
+	log.Debug().Msg("reaping the mempool")
+	txMempool := uow.GetBus().GetUtilityModule().GetMempool()
+	if txs, err = uow.reapMempool(txMempool, maxTxBytes); err != nil {
+		return "", nil, err
+	}
+
+	// end block lifecycle phase
+	log.Debug().Msg("calling endBlock")
+	if err := uow.endBlock(proposer); err != nil {
 		return "", nil, err
 	}
+
+	log.Debug().Msg("computing state hash")
+	// TODO(@deblasis): this should be from a ReadContext (the ephemeral/staging one)
```

Comment:
We need to get the state hash from the modified context (i.e. the one we're writing to), so I believe this is correct and we can remove the TODO